### PR TITLE
VtGateV3 sequence testcases in GO

### PR DESCRIPTION
**Pull request by ajeetj**
Rewrote VtGateV3 sequence testcases from Python to Go. 
Related python code reference https://github.com/planetscale/vitess/blob/master/test/vtgatev3_test.py#L1575-L1609

This is the first PR for review and to collect feedback. 

Note: We all are merging changes related to vtgate in "tal_vtgate" repo.

### CODE REVIEW

File: go/vt/vtgate/endtoend/sequence_test.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+func TestSeq(t *testing.T) {
+	ctx := context.Background()
+	conn, err := mysql.Connect(ctx, &vtParams)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer conn.Close()
+
+	//Why below query doesn't work for seq table?
+	//exec(t, conn, "insert into sequence_test_seq(id, next_id, cache) values(0,1,10)")
```
Comment (ajeetj) [inline go/vt/vtgate/endtoend/sequence_test.go:36]: @deepthi  Do you know why the above line will not work? 
I had to pass this insert as a setup DDL to make it work.

File: go/vt/vtgate/endtoend/main_test.go [hunk 2]
```diff
@@ -152,6 +167,33 @@ create table t2_id4_idx(
...
+				Pinned: "80",
+				Columns: []*vschemapb.Column{{
+					Name: "id",
+					// Type: sqltypes.Int64,
+				}, {
+					Name: "next_id",
+					// Type: sqltypes.Int64,
+				}, {
+					Name: "cache",
+					// Type: sqltypes.Int64,
```
Comment (ajeetj) [inline go/vt/vtgate/endtoend/main_test.go:212]: @deepthi Do we need to list all the columns from a sequence table in VSchema?

File: go/vt/vtgate/endtoend/sequence_test.go (hunk 1 above)
Comment (deepthi) [reply go/vt/vtgate/endtoend/sequence_test.go:36]: What error did you get?

Comment (ajeetj) [reply go/vt/vtgate/endtoend/sequence_test.go:36]: I get below error: Is there any restriction that the seq table should be prepopulated before bringing the cluster? 

```
E1007 11:53:54.143075   17761 server.go:270] mysql_server caught panic:
runtime error: index out of range [0] with length 0
/usr/local/go/src/runtime/panic.go:75 (0x102f322)
	goPanicIndex: panic(boundsError{x: int64(x), signed: true, y: y, code: boundsIndex})
/Users/ajeetjain/ps/vitess/go/vt/vtgate/engine/insert.go:394 (0x1a27ca0)
	io/vitess/go/vt/vtgate/engine.(*Insert).getInsertShardedRoute: keyspaceIDs, err := ins.processPrimary(vcursor, vindexRowsValues[0], ins.Table.ColumnVindexes[0], bindVars)
/Users/ajeetjain/ps/vitess/go/vt/vtgate/engine/insert.go:272 (0x1a25488)
	io/vitess/go/vt/vtgate/engine.(*Insert).execInsertSharded: rss, queries, err := ins.getInsertShardedRoute(vcursor, bindVars)
/Users/ajeetjain/ps/vitess/go/vt/vtgate/engine/insert.go:222 (0x1a249f2)
	io/vitess/go/vt/vtgate/engine.(*Insert).Execute: return ins.execInsertSharded(vcursor, bindVars)
/Users/ajeetjain/ps/vitess/go/vt/vtgate/executor.go:304 (0x1a780c8)
	io/vitess/go/vt/vtgate.(*Executor).handleExec: qr, err := plan.Instructions.Execute(vcursor, bindVars, true)
/Users/ajeetjain/ps/vitess/go/vt/vtgate/executor.go:210 (0x1a769f7)
	io/vitess/go/vt/vtgate.(*Executor).execute: qr, err := e.handleExec(ctx, safeSession, sql, bindVars, destKeyspace, destTabletType, dest, logStats)
/Users/ajeetjain/ps/vitess/go/vt/vtgate/executor.go:134 (0x1a7659b)
	io/vitess/go/vt/vtgate.(*Executor).Execute: result, err = e.execute(ctx, safeSession, sql, bindVars, logStats)
/Users/ajeetjain/ps/vitess/go/vt/vtgate/vtgate.go:289 (0x1a9a741)
	io/vitess/go/vt/vtgate.(*VTGate).Execute: qr, err = vtg.executor.Execute(ctx, "Execute", NewSafeSession(session), sql, bindVariables)
/Users/ajeetjain/ps/vitess/go/vt/vtgate/plugin_mysql_server.go:206 (0x1a883d5)
	io/vitess/go/vt/vtgate.(*vtgateHandler).ComQuery: session, result, err := vh.vtg.Execute(ctx, session, query, make(map[string]*querypb.BindVariable))
/Users/ajeetjain/ps/vitess/go/mysql/conn.go:1070 (0x148e647)
	io/vitess/go/mysql.(*Conn).execQuery: err := handler.ComQuery(c, query, func(qr *sqltypes.Result) error {
/Users/ajeetjain/ps/vitess/go/mysql/conn.go:768 (0x148dec0)
	io/vitess/go/mysql.(*Conn).handleNextCommand: if err := c.execQuery(sql, handler, more); err != nil {
/Users/ajeetjain/ps/vitess/go/mysql/server.go:452 (0x14a8e0e)
	io/vitess/go/mysql.(*Listener).handle: err := c.handleNextCommand(l.handler)
/usr/local/go/src/runtime/asm_amd64.s:1357 (0x1060b30)
	goexit: BYTE	$0x90	// NOP
--- FAIL: TestSeq (0.01s)
    sequence_test.go:36: EOF (errno 2013) (sqlstate HY000) during query: insert into sequence_test_seq(id, next_id, cache) values(0,1,10)
I1007 11:53:54.149751   17761 engine.go:278]
```

**Comment by ajeetj** [comment]
@deepthi The above tests runs fine for 1st and 2nd time but fails if I keep executing them multiple times. I am not able to figure out why is that. I tried to use transaction begin/commit or sync.WaitGroup or the sleep function but saw the same behavior. 
Could this be because I am creating the sequence table in a sharded keyspace?

**Comment by deepthi** [comment]
> @deepthi The above tests runs fine for 1st and 2nd time but fails if I keep executing them multiple times. I am not able to figure out why is that. I tried to use transaction begin/commit or sync.WaitGroup or the sleep function but saw the same behavior.
> Could this be because I am creating the sequence table in a sharded keyspace?

This means that the test's tearDown is not cleaning things up properly.
But you bring up a good point. Why is the sequence table in the sharded keyspace? Is that what the original python test was doing? This might be the cause of the panic you were getting.

**Comment by ajeetj** [comment]
I couldn't find a way to add multiple keyspaces in the current go
implementation using the VTCOMBO. And then I noticed in the documents that
it is possible to put the sequence table in a sharded keyspace by pinning
it to a single shard. So I was trying with that.
Teardown is working fine as first test is always passing. Its the test 2
and 3 which fails intermittently, I doubt if I am setting up them wrong.

**Comment by deepthi** [comment]
> I couldn't find a way to add multiple keyspaces in the current go implementation using the VTCOMBO. And then I noticed in the documents that it is possible to put the sequence table in a sharded keyspace by pinning it to a single shard. So I was trying with that. Teardown is working fine as first test is always passing. Its the test 2 and 3 which fails intermittently, I doubt if I am setting up them wrong.

I see that VTTestTopology does allow you to have multiple keyspaces. If you were to add another unsharded keyspace to the list it has, and separate the schema and vschema for sequence_test_seq, is that causing a problem later on?

File: go/vt/vtgate/endtoend/main_test.go (hunk 2 above)
Comment (deepthi) [reply go/vt/vtgate/endtoend/main_test.go:212]: I don't think so. Something like this should work (in an unsharded keyspace):
```
{
    "tables": {
        "test_seq": {
            "type": "sequence"
        }
     }
}
```

Comment (ajeetj) [reply go/vt/vtgate/endtoend/main_test.go:212]: Yes, I tested that yesterday it worked. I don't remember the error I was getting before.

**Comment by ajeetj** [comment]
> > I couldn't find a way to add multiple keyspaces in the current go implementation using the VTCOMBO. And then I noticed in the documents that it is possible to put the sequence table in a sharded keyspace by pinning it to a single shard. So I was trying with that. Teardown is working fine as first test is always passing. Its the test 2 and 3 which fails intermittently, I doubt if I am setting up them wrong.
> 
> I see that VTTestTopology does allow you to have multiple keyspaces. If you were to add another unsharded keyspace to the list it has, and separate the schema and vschema for sequence_test_seq, is that causing a problem later on?

Yes, I tried that as well yesterday but was stuck in the config.SchemaDir. It complains that it has already set. I also notice below comment on the "InitSchemas" function
```
// InitSchemas is a shortcut for tests that just want to setup a single
// keyspace with a single SQL file, and/or a vschema.
// It creates a temporary directory, and puts the schema/vschema in there.
// It then sets the right value for cfg.SchemaDir.
// At the end of the test, the caller should os.RemoveAll(cfg.SchemaDir).
func (cfg *Config) InitSchemas(keyspace, schema string, vschema *vschemapb.Keyspace) error {
```
I couldn't find any other implementation which uses multiple VSchema for GO endtoend testing.  Maybe we will need to priorities writing the new cluster code now.

File: go/vt/vtgate/endtoend/sequence_test.go (hunk 1 above)
Comment (deepthi) [reply go/vt/vtgate/endtoend/sequence_test.go:36]: This is because sequence_test_seq is in a sharded keyspace, but no vindex is defined for it.

**Comment by deepthi** [comment]
Can you try either:
a) change InitSchemas to work for multiple keyspaces?
or
b) write another function that will do that?

The preference is (a)

**Comment by deepthi** [comment]
> Teardown is working fine as first test is always passing. Its the test 2 and 3 which fails intermittently, I doubt if I am setting up them wrong.

I see what you mean. First run passes and then it fails either on `[[INT64(5)]]` or `[[INT64(11)]]`
It is a mystery to me right now. Maybe fixing the keyspace / schema creation will help.

**Comment by ajeetj** [comment]
> Can you try either:
> a) change InitSchemas to work for multiple keyspaces?
> or
> b) write another function that will do that?
> 
> The preference is (a)

OK, Will look into this today. Thanks

**Comment by ajeetj** [comment]
Rewrote the above using a new cluster which is working fine. 
Ref: https://github.com/planetscale/vitess/pull/3

@deepthi 
Do we need to debug why this was failing?

**Comment by deepthi** [comment]
> Rewrote the above using a new cluster which is working fine.
> Ref: #3
> 
> @deepthi
> Do we need to debug why this was failing?

Not needed.

**Pull request closed without merging**